### ISSUE TITLE
Feature Request: Recipe Step(s) to compute group-meaned and de-meaned variables

### ISSUE THREAD

**Issue by AndrewKostandy** [tidymodels/recipes#931]
Now that {multilevelmod} is on CRAN, it would be great if we can have one or more recipe steps that compute group-meaned and de-meaned variables like the new {datawizard} functions ```demean()```, ```degroup()```, and ```detrend()``` do.

This will be useful when we want to model the between and within subject effects.

For more info about the use of this functionality, please see the documentation for ```demean()``` by running ```?datawizard::demean``` .

Thank you for all your work!

**Comment by AndrewKostandy** [comment]
So this is how it would be done with dplyr for one variable. I think there could be one recipe step for each of the mutate functions below.

Yes, `step_center()` with a grouping variable could replace the 2nd mutate and we'd need to introduce the mean (create the "_between" variable) in a previous step (the 1st mutate) before the variable is removed by `step_center()`.

```r
library(dplyr)

iris_demeaned <- iris %>% 
  select(Petal.Length, Species) %>% 
  group_by(Species) %>%
  mutate(Petal.Length_between = mean(Petal.Length)) %>% # recipe step 1 with group argument
  mutate(Petal.Length_within = Petal.Length - mean(Petal.Length), .keep = "unused") %>% # recipe step 2 with group argument
  ungroup()

iris_demeaned
#> # A tibble: 150 × 3
#>    Species Petal.Length_between Petal.Length_within
#>    <fct>                  <dbl>               <dbl>
#>  1 setosa                  1.46             -0.0620
#>  2 setosa                  1.46             -0.0620
#>  3 setosa                  1.46             -0.162 
#>  4 setosa                  1.46              0.0380
#>  5 setosa                  1.46             -0.0620
#>  6 setosa                  1.46              0.238 
#>  7 setosa                  1.46             -0.0620
#>  8 setosa                  1.46              0.0380
#>  9 setosa                  1.46             -0.0620
#> 10 setosa                  1.46              0.0380
#> # … with 140 more rows
```

**Comment by EmilHvitfeldt** [comment]
Out of curiosity. Would there be a need for 2 steps or could be combine into one step? rephrased differently, would there be a time where you would want to calculate the `_between` variable without the `_within` variable?

**Comment by AndrewKostandy** [comment]
Yes there would need to be 2 steps because there are various scenarios you may want to do:

1) You may want to keep the variable uncentered and introduce the per context/cluster mean
In that case I would rename the variable itself as _within while calling the per context/cluster mean as _contextual as it would represent the contextual effect.

```r
iris %>% 
  select(Petal.Length, Species) %>% 
  rename(Petal.Length_within = Petal.Length) %>%
  group_by(Species) %>%
  mutate(Petal.Length_contextual = mean(Petal.Length_within)) %>% 
  ungroup() 
```
2) You may want to center the variable using the grand mean and introduce the per context/cluster mean. In that case I would call the grand-mean centered variable as _within while calling the per context/cluster mean as _contextual as it would represent the contextual effect.

```r
iris %>% 
  select(Petal.Length, Species) %>% 
  mutate(Petal.Length_within = Petal.Length - mean(Petal.Length), .keep = "unused") %>% # Uses the grand mean for centering
  group_by(Species) %>%
  mutate(Petal.Length_contextual = mean(Petal.Length_within)) %>% 
  ungroup() 
```

3)  The scenario shown in my earlier post which centers within context/cluster and introduces the per context/cluster mean. I slightly reordered the steps below to look similar to the above scenarios. In this case the variable centered within context/cluster would be called _within while the per context/cluster mean would be called _between as it would represent the between effect.

```r
iris %>% 
  select(Petal.Length, Species) %>% 
  group_by(Species) %>%
  mutate(Petal.Length_within = Petal.Length - mean(Petal.Length)) %>% 
  mutate(Petal.Length_between = mean(Petal.Length),  .keep = "unused") %>%
  ungroup()
```

**Comment by topepo** [comment]
TBH I'm worried that this could be misused. For example, grouping on the outcome class etc. 

I think that it's a good idea that should live in an extension package.
